ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -535,3 +555,77 @@ const char* VirtualizationRoot_GetRootRelativePath(VirtualizationRootHandle root
     
     return relativePath;
 }
+
+static bool UsedMountPoint_MountEquals(const mount_t& mountPoint, const UsedMountPoint& usage)
+{
+    return mountPoint == usage.mountPoint;
+}
+
+static void IncrementAuthCacheDisableCount(mount_t mountPoint)
+{
+    RWLock_AcquireExclusive(s_virtualizationRootsLock);
+    {
+        UsedMountPoint* usedMountPoint = s_usedMountPoints.FindOrInsertLockedAllowResize(
+            mountPoint,
+            UsedMountPoint { .mountPoint = mountPoint },
+            s_virtualizationRootsLock);
+    
+        if (nullptr == usedMountPoint)
+        {
+            // Alloc failure, give up trying to track the mount point
+            KextLog_Note("IncrementAuthCacheDisableCount: failed to resize array, disabling auth cache on mount point %s without saving previous value", vfs_statfs(mountPoint)->f_mntonname);
+            vfs_setauthcache_ttl(mountPoint, 0);
+        }
+        else
+        {
+            if (usedMountPoint->authCacheDisableCount == 0)
+            {
+                usedMountPoint->savedAuthCacheTTL = vfs_authcache_ttl(mountPoint);
+                vfs_setauthcache_ttl(mountPoint, 0);
+                KextLog_Info("IncrementAuthCacheDisableCount: disabling auth cache on mount point '%s' saved TTL = %d",
+                    vfs_statfs(mountPoint)->f_mntonname, usedMountPoint->savedAuthCacheTTL);
+            }
+       
+            usedMountPoint->authCacheDisableCount++;
+        }
+    }
+    RWLock_ReleaseExclusive(s_virtualizationRootsLock);
+}
+
+static void DecrementAuthCacheDisableCount_Locked(mount_t mountPoint)
```

Comment:
Is someone force kills a mount (i.e. no unmount), is there a way to detect and call this?

Reply:
That's already covered; in this situation, the [`IOUserClient::clientDied()` function is called, which calls `IOUserClient::clientClose()`](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/iokit/Kernel/IOUserClient.cpp#L1555). Our provider user client class [overrides that function](https://github.com/Microsoft/VFSForGit/blob/0459b281d2ec259f0749157e2924bb0ec0580f37/ProjFS.Mac/PrjFSKext/PrjFSKext/PrjFSProviderUserClient.cpp#L95), which eventually calls down to the auth cache decrement via `ActiveProvider_Disconnect()`.

Reply:
Thanks for that explanation makes sense!